### DISCUSSION TITLE
Allow `NULL` to be legit target result, store `R_MissingArg` instead

### DISCUSSION THREAD

**Discussion by dipterix** [ropensci/targets#858] [General]
### Issue

Currently nodes returning `NULL` will be considered "invalid":

> "null": The errored target continues and returns NULL. The data hash is deliberately wrong so the target is not up to date for the next run of the pipeline.

As a result, even if **the underlying result should be `NULL`**, the pipeline will still consider the node to be invalid. One of the consequences I often find is when making targets with `shortcut=TRUE`, the pipeline raises errors if `NULL` values are encountered.

Is there any way to avoid raising errors in such cases? Maybe users can set options to ignore NULL... but then I found out if a node is truly invalid, then we still need to raise errors.... Duh

### Propose

Maybe we can consider a less frequently used symbol (you can create it via `alist(a=)[[1]]`) to store "invalid results"? This symbol is very special, normally you are unlikely to create it elsewhere in base R (for example, `str2lang("")` and `quote()` will raise errors) unless using the low-level `C` interface via `R_MissingArg` symbol. Also trying to evaluate this symbol will trigger "missing" error (instead of "object not found"):

```r
missing_arg <- alist(a=)[[1]]
eval(missing_arg)
#> Error in eval(missing_arg) : 
#>   argument "missing_arg" is missing, with no default
```

This symbol perfectly fits in the situation where a target is errored (the corresponding value is missing!).

One propose I want to make is to 

1. consider both `NULL` and `R_MissingArg` symbols as invalid values by default, so current users won't be affected
2. store `R_MissingArg` as "invalid" node values instead of `NULL`. The hash values of errored nodes will change, but the value is out-dated anyway, so...
3. provide options to treat `NULL` as valid. This can be enabled by default, or just in shortcut mode.

If you don't like `alist(a=)[[1]]`, I think `rlang::sym("")` also generates the symbol.

**Comment by wlandau** [comment]
That excerpt from the documentation is about what happens if you set the `error` argument to `"null"` in `tar_target()` or `tar_option_set()`. All it means is that if `error` is `"null"` and the target throws an error, then the return value will be `NULL` and the pipeline will continue. `NULL` is still a legitimate result for a target that does not throw an error.

**Reply by dipterix** [reply]
Gotcha. Thanks so much for explaining this to me. 

So from what I understand, when a result is `NULL`, I should still be able to use`tar_make` with `shortcut=TRUE`?

However, I encountered errors when some target values are `NULL` and shortcut wouldn't allow me to bypass those targets.

**Reply by wlandau** [reply]
Okay, I think see now. 5294be93db961a492747cdcc81f85a09648b90d5 should fix that problem.

**Reply by dipterix** [reply]
Hooray! 

Looks like these two are different problems (shortcut failure and using `NULL` as default value for errored targets). Still worth to consider `R_MissingArg`, but I guess it's an edge case now.

**Reply by wlandau** [reply]
I like NULL because it is well understood to be the absence of a value and because user-side code is more likely to be able to handle it in downstream targets. And I would prefer not to have any custom compiled code in the package.
